app/controllers/application_controller/compare.rb

```diff
@@ -412,14 +412,60 @@ def sections_field_changed
     end
   end
 
-  def set_checked_sections
-    session[:selected_sections] = []
-    params[:all_checked]&.each do |a|
-      s = a.split(':')
-      if s.length > 1
-        session[:selected_sections].push(s[1])
+  # Method to get the selection names from :all_checked and :id params.
+  # individual selected item suggests that the variable name contains ':'
+  # else, the child_names are retrived from the selected_parent item.
+  def selection_names(name)
+    split_name = name.split(':')
+    split_name.length > 1 ? [split_name[1]] : child_names(name)
+  end
+
+  # Method to get the child names of a selected parent item.
+  # session[:miq_sections] contains all tree names and the keys are filters using the selected group_name[1].
+  def child_names(name)
+    child_names = []
+    group_name = name.split('_')
+    if group_name.length == 1
+      child_sections = session[:miq_sections]&.select { |_key, value| value[:group] == group_name[1] }
+      if child_sections && child_sections.length > 1
+        child_names = child_sections.keys.map(&:to_s)
       end
     end
+    child_names
+  end
+
+  # Method to add or remove selections if an id is present.
+  def update_selections(param_id, selections, check)
+    if param_id
+      name = selection_names(CGI.unescape(param_id))
+      check == 'true' ? add_selection(name, selections) : remove_selections(name, selections)
+    else
+      selections
+    end
+  end
+
+  # Method to add names to the selections.
+  def add_selection(names, selections)
+    names.each do |i|
+      selections.push(i) if selections.exclude?(i)
+    end
+    selections
+  end
+
+  # Method to remove names to the selections.
+  def remove_selections(names, selections)
+    names.each do |i|
+      selections.delete(i) if selections.include?(i)
+    end
+    selections
+  end
+
+  def set_checked_sections
+    session[:selected_sections] = selections = []
+    params[:all_checked]&.each do |item|
```

Comment:
Do we need some sort of parameter validation here?  What if a user sends in some garbage or hacky variables or something?